oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

```diff
@@ -581,94 +498,295 @@ private MongoFilterPaths getPathsForRegexFiltering() {
+                        numberOfFailures++;
+                        LOG.warn("Retrying download in {} ms; number of times failed: {}; current series of failures started at: {} ({} seconds ago)",
+                                retryIntervalMs, numberOfFailures, failuresStartTimestamp, secondsSinceStartOfFailures);
+                        exceptions.compute(e.getClass().getSimpleName() + " - " + e.getMessage(),
+                                (key, val) -> val == null ? 1 : val + 1
+                        );
+                        Thread.sleep(retryIntervalMs);
+                        // simple exponential backoff mechanism
+                        retryIntervalMs = Math.min(retryMaxIntervalMillis, retryIntervalMs * 2);
                     }
                 }
-                if (nextIndex > 0) {
-                    LOG.info("Enqueueing last block with {} elements, estimated size: {}",
-                            nextIndex, IOUtils.humanReadableByteCountBin(batchSize));
-                    tryEnqueueCopy(batch, nextIndex);
+            }
+        }
+
+        private void downloadRange(DownloadRange range, Bson filter, DownloadOrder downloadOrder) throws InterruptedException, TimeoutException {
+            Bson findQuery = range.getFindQuery();
+            if (filter != null) {
+                findQuery = Filters.and(findQuery, filter);
+            }
+            Bson sortOrder = downloadOrder.downloadInAscendingOrder() ?
+                    Sorts.ascending(NodeDocument.MODIFIED_IN_SECS, NodeDocument.ID) :
+                    Sorts.descending(NodeDocument.MODIFIED_IN_SECS, NodeDocument.ID);
+
+            FindIterable<NodeDocument> mongoIterable = dbCollection
+                    .find(findQuery)
+                    .sort(sortOrder);
+
+            LOG.info("Traversing: {}. Query: {}, Traverse order: {}", range, findQuery, sortOrder);
+            download(mongoIterable);
+        }
+
+        void download(FindIterable<NodeDocument> mongoIterable) throws InterruptedException, TimeoutException {
+            try (MongoCursor<NodeDocument> cursor = mongoIterable.iterator()) {
+                NodeDocument[] batch = new NodeDocument[maxBatchNumberOfDocuments];
+                int nextIndex = 0;
+                int batchSize = 0;
+                if (cursor.hasNext()) {
+                    // We have managed to reconnect, reset the failure timestamp
+                    failuresStartTimestamp = null;
+                    numberOfFailures = 0;
                 }
-            } catch (MongoException e) {
-                if (e instanceof MongoInterruptedException || e instanceof MongoIncompatibleDriverException) {
-                    // Non-recoverable exceptions
+                try {
+                    while (cursor.hasNext()) {
+                        NodeDocument next = cursor.next();
+                        String id = next.getId();
+                        this.nextLastModified = next.getModified();
```

Comment:
My editor warns me that this might produce a `NullPointerException`. And we use this a few other places.

Reply:
Good catch, this would indeed be a problem when doing a column traversal because the query used for that was downloading all documents, even those without the `_modified` field. We do not need these documents to build the FFS, so it is safe to filter them on the Mongo query. Like this, there is not need to have a null check here. This code is on the critical path, so it should be kept as lean as possible.